library/std/src/sys/unix/weak.rs

```diff
@@ -84,18 +84,26 @@ macro_rules! syscall {
     )
 }
 
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 macro_rules! syscall {
     (fn $name:ident($($arg_name:ident: $t:ty),*) -> $ret:ty) => (
         unsafe fn $name($($arg_name:$t),*) -> $ret {
             // This looks like a hack, but concat_idents only accepts idents
             // (not paths).
             use libc::*;
 
-            syscall(
-                concat_idents!(SYS_, $name),
-                $($arg_name as c_long),*
-            ) as $ret
+            weak! { fn $name($($t),*) -> $ret }
+
+            // Use a weak symbol from libc when possible, allowing `LD_PRELOAD`
+            // interposition, but if it's not found just use a raw syscall.
+            if let Some(fun) = $name.get() {
+                fun($($arg_name),*)
+            } else {
+                syscall(
+                    concat_idents!(SYS_, $name),
+                    $($arg_name as c_long),*
```

Comment:
This is probably where all arguments are made ptr-sized. It is not clear to me what that is done, nor why it should even be allowed. Isn't it UB when the caller and callee types do not match?

Reply:
This is probably to make sure they're all integral/pointer values that fit in a machine word. As long as that's the case, the exact type doesn't matter. The ABI of a variadic function like `syscall` will just put each argument into their own register, extended to a `usize`/`c_long`.

Reply:
Is this a guarantee made by C? For *all* ABIs? And what happens with the value that is passed, logically, to the syscall? It should get truncated to the smaller of "type used" and "type expected", I guess?

Reply:
I propose we do something like https://github.com/rust-lang/rust/pull/79196.